src/v/kafka/protocol/schemata/generator.py

```diff
@@ -1444,6 +1455,60 @@ class response;
 {%- endif %}
 {%- endmacro %}
 
+{% macro render_errored_source(struct) %}
+{%- if op_type == "response" %}
+{%- if not struct.fields %}
+ss::future<bool> {{ struct.name }}::errored() const {
+    co_return false;
+}
+{%- else %}
+ss::future<bool> {{ struct.name }}::errored() const {
+    bool ret = false;
+    {%- for field in struct.fields %}
+    {%- if field.is_error_code %}
+    if ({{ field.name }} != kafka::error_code::none) {
+        co_return true;
+    }
+    {%- elif field.is_array and field.type().value_type().is_struct %}
+    {%- if field.nullable() %}
+    if ({{ field.name }}) {
+        ret = co_await ss::map_reduce(
+          {{ field.name }}->begin(),
+          {{ field.name }}->end(),
+          [](auto& item) { return item.errored(); },
+          bool(false),
+          [](bool result, bool val) {
+              return result || val;
+          });
```

Comment:
I'm not sure why this is async? Are some responses so large that checking for an error creates reactor stalls?

I haven't compiled this, but the synchronous version is more like:
```suggestion
        ret = std::any_of(
          {{ field.name }}->begin(),
          {{ field.name }}->end(),
          [](auto& item) { return item.errored(); });
```

Reply:
> I'm not sure why this is async? Are some responses so large that checking for an error creates reactor stalls?

Correct. I am thinking about production systems where some responses, such as a produce, could be quite large. So iterating through the response hierarchy sequentially would be slow. Thus, the async version.

Reply:
Is this a real problem?

Adding asynchrony to the logging here means that log entries may be reordered, which may make it significantly harder to reason about errors, since we don't have a request id.

Reply:
> Adding asynchrony to the logging here means that log entries may be reordered, which may make it significantly harder to reason about errors, since we don't have a request id.

The response is not modified here so I do not see how log entries are reordered. The asynchrony here is simply to search for a Kafka error.

Reply:
> I do not see how log entries are reordered.

The suspension points mean a single response log could be delayed arbitrarily.

Reply:
Or is the concern here about when message is logged (i.e., when we call `klog.debug()`)? If yes, this seems like a non-concern since the original `ctx.respond()` is already async.

Reply:
Changed to sync version.